src/main/java/pl/psnc/dei/service/PersistableExceptionService.java

```diff
@@ -0,0 +1,127 @@
+import pl.psnc.dei.model.Record;
+import pl.psnc.dei.model.conversion.Context;
+import pl.psnc.dei.model.exception.TranscriptionPlatformException;
+import pl.psnc.dei.service.context.ContextMediator;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class PersistableExceptionService {
+
+    @Autowired
+    private PersistableExceptionRepository persistableExceptionRepository;
+
+    @Autowired
+    private ContextMediator contextMediator;
+
+    public <T extends  Exception> T findByContextAndExceptionClass(Context context, Class<T> exceptionClass) {
+        PersistableException.ExceptionType exceptionType = this.convertExceptionClassToExceptionType(exceptionClass);
+        Optional<PersistableException> optionalExceptions = this.persistableExceptionRepository.findByContextAndType(context, exceptionType);
+        return optionalExceptions.map(persistableException -> this.inflateException(persistableException, exceptionClass)).orElse(null);
+    }
+
+    public PersistableException bind(Exception exception, Context context) {
+        PersistableException.ExceptionType exceptionType = this.convertExceptionClassToExceptionType(exception.getClass());
+        Optional<PersistableException> optionalException = this.persistableExceptionRepository.findByContextAndType(context, exceptionType);
+        if (optionalException.isPresent()) {
+            PersistableException fetchedException = optionalException.get();
+            fetchedException.setMessage(exception.getMessage());
+            PersistableException persistableException = this.persistableExceptionRepository.save(fetchedException);
+            context.getExceptions().add(persistableException);
+            this.contextMediator.save(context);
+            return persistableException;
+        }
+        else {
+            PersistableException newPersistableException = new PersistableException();
+            newPersistableException.setContext(context);
+            newPersistableException.setMessage(exception.getMessage());
+            newPersistableException.setType(this.convertExceptionClassToExceptionType(exception.getClass()));
+            newPersistableException = this.persistableExceptionRepository.save(newPersistableException);
+            context.getExceptions().add(newPersistableException);
+            this.contextMediator.save(context);
+            return newPersistableException;
+        }
+    }
+
+    private <T extends Exception> T inflateException(PersistableException persistableException, Class<T> exceptionClass) {
+        switch (persistableException.getType()) {
+            case TRANSCRIPTION_PLATFORM_EXCEPTION: {
+                // unchecked cast without it
+                if (exceptionClass.isAssignableFrom(TranscriptionPlatformException.class)) {
+                    return exceptionClass.cast(new TranscriptionPlatformException(persistableException.getMessage()));
+                }
+            }
```

Comment:
There are no break instructions. Is this on purpose? It would be strange because it would mean ignoring exception type...

Reply:
breaks was removed to shrink code. Each `PersistableException` stores type of exception it was, thus it will be matched on first try to right case in which method returns value and stops switch execution. If statement is used to get rid off warning about unchecked cast